controllers/spacebindingrequest/spacebindingrequest_controller_test.go

```diff
@@ -79,6 +123,518 @@ func TestCreateSpaceBindingRequest(t *testing.T) {
+			spaceBinding.DeletionTimestamp = &metav1.Time{Time: time.Now().Add(-121 * time.Second)} // is being deleted since more than 2 minutes
+			member1 := NewMemberClusterWithClient(test.NewFakeClient(t, sbr, sbrNamespace), "member-1", corev1.ConditionTrue)
+			hostClient := test.NewFakeClient(t, base1nsTier, janeSpace, janeMur, spaceBinding)
+			ctrl := newReconciler(t, hostClient, member1)
+
+			// when
+			_, err := ctrl.Reconcile(context.TODO(), requestFor(sbr))
+
+			// then
+			cause := "spacebinding deletion has not completed in over 2 minutes"
+			require.EqualError(t, err, cause)
+			spacebindingrequesttest.AssertThatSpaceBindingRequest(t, sbr.GetNamespace(), sbr.GetName(), member1.Client).
+				HasConditions(spacebindingtest.TerminatingFailed(cause)).
+				HasFinalizer()
+		})
+
+		t.Run("unable to delete SpaceBinding", func(t *testing.T) {
+			// given
+			sbr := spacebindingrequesttest.NewSpaceBindingRequest("jane", sbrNamespace.GetName(),
+				spacebindingrequesttest.WithDeletionTimestamp(),
+				spacebindingrequesttest.WithFinalizer(),
+			) // sbr is being deleted
+			spaceBinding := spacebindingtest.NewSpaceBinding(janeMur.Name, janeSpace.Name, "admin", sbr.Name, spacebindingtest.WithSpaceBindingRequest(sbr))
+			member1 := NewMemberClusterWithClient(test.NewFakeClient(t, sbr, sbrNamespace), "member-1", corev1.ConditionTrue)
+			hostClient := test.NewFakeClient(t, base1nsTier, janeSpace, janeMur, spaceBinding)
+			hostClient.MockDelete = func(ctx context.Context, obj runtimeclient.Object, opts ...runtimeclient.DeleteOption) error {
+				if _, ok := obj.(*toolchainv1alpha1.SpaceBinding); ok {
+					return fmt.Errorf("mock error")
+				}
+				return hostClient.Client.Delete(ctx, obj, opts...)
+			}
+			ctrl := newReconciler(t, hostClient, member1)
+
+			// when
+			_, err := ctrl.Reconcile(context.TODO(), requestFor(sbr))
+
+			// then
+			cause := "unable to delete spacebinding: mock error"
+			require.EqualError(t, err, cause)
+			spacebindingrequesttest.AssertThatSpaceBindingRequest(t, sbr.GetNamespace(), sbr.GetName(), member1.Client).
+				HasConditions(spacebindingtest.TerminatingFailed(cause)).
+				HasFinalizer()
+		})
+
+		t.Run("failed to remove finalizer", func(t *testing.T) {
+			// given
+			member1Client := test.NewFakeClient(t, sbr, sbrNamespace)
+			member1Client.MockUpdate = func(ctx context.Context, obj runtimeclient.Object, opts ...runtimeclient.UpdateOption) error {
+				if _, ok := obj.(*toolchainv1alpha1.SpaceBindingRequest); ok {
+					return fmt.Errorf("mock error")
+				}
+				return member1Client.Client.Update(ctx, obj, opts...)
+			}
+			member1 := NewMemberClusterWithClient(member1Client, "member-1", corev1.ConditionTrue)
+			hostClient := test.NewFakeClient(t)
+			ctrl := newReconciler(t, hostClient, member1)
+
+			// when
+			_, err := ctrl.Reconcile(context.TODO(), requestFor(sbr))
+
+			// then
+			require.EqualError(t, err, "failed to remove finalizer: mock error")
+		})
```

Comment:
great work on testing and mocking this :D +1

Reply:
I can't take credit for this , I was just good at copy pasting and following same approach I saw in other unit tests.
The repository mature and full of great examples on how to solve many of these "tricky" problems with testing and not only. 

Kudos to the team 🏅